new arithmetic overflow rule (#220)
rustfoundation/safety-critical-rust-coding-guidelines: src/coding-guidelines/expressions.rst

Expressions
===========

.. guideline:: Ensure that integer operations do not result in arithmetic overflow
    :id: gui_dCquvqE1csI3
    :category: required
    :status: draft
    :release: 1.0 - latest
    :fls: fls_oFIRXBPXu6Zv
    :decidability: decideable
    :scope: system
    :tags: security, performance, numerics

    Eliminate `arithmetic overflow <https://rust-lang.github.io/fls/expressions.html#arithmetic-overflow>`_ of both signed and unsigned integer types. 
    Arithmetic overflow will panic in debug mode, but wraparound in release mode.
    Any wraparound behavior must be explicitly specified to ensure the same behavior in both debug and release modes.

    This rule applies to the following primitive types:

    * ``i8``
    * ``i16``
    * ``i32``
    * ``i64``
    * ``i128``
    * ``u8``
    * ``u16``
    * ``u32``
    * ``u64``
    * ``u128``
    * ``usize``
    * ``isize``

    .. rationale::
        :id: rat_LvrS1jTCXEOk
        :status: draft

        Eliminate arithmetic overflow to avoid runtime panics and unexpected wraparound behavior.
        Use explicit wrapping or saturation semantics in cases where these behaviors are intentional.

>         Range checking can be used to eliminate the possibility of arithmetic overflow.

Reviewer: Here I would add the line that I just "suggestion-deleted" from the Amplification. Then adjust the rest of the paragraph to not repeat what you would've just said (since that line would now be here in the Rationale).

	        fn add(si_a: i32, si_b: i32) -> Result<i32, ArithmeticError> {
                if (si_b > 0 && si_a > INT_MAX - si_b)
                    || (si_b < 0 && si_a < INT_MIN - si_b)
                {
                    Err(ArithmeticError::Overflow)
                } else {
                    Ok(si_a + si_b)
                }
            }

            fn sub(si_a: i32, si_b: i32) -> Result<i32, ArithmeticError> {
                if (si_b < 0 && si_a > INT_MAX + si_b)
                    || (si_b > 0 && si_a < INT_MIN + si_b)
                {
                   Err(ArithmeticError::Overflow)
                } else {
                   Ok(si_a - si_b)
                }
            }

	        fn mul(si_a: i32, si_b: i32) -> Result<i32, ArithmeticError> {
                if si_a == 0 || si_b == 0 {
                    return Ok(0);
                }

                // Detect overflow before performing multiplication
                if (si_a == -1 && si_b == INT_MIN) || (si_b == -1 && si_a == INT_MIN) {
                    Err(ArithmeticError::Overflow)
                } else if (si_a > 0 && (si_b > INT_MAX / si_a || si_b < INT_MIN / si_a))

>                     || (si_a < 0 && (si_b > INT_MIN / si_a || si_b < INT_MAX / si_a))

Author: done

                {
                    Err(ArithmeticError::Overflow)
                } else {
                    Ok(si_a * si_b)
                }

>             }

Author: yes, I see it.

        .. code-block:: rust

            fn add(si_a: i32, si_b: i32) -> Result<i32, ArithmeticError> {
                si_a.checked_add(si_b).ok_or(ArithmeticError::Overflow)
            }

            fn sub(a: i32, b: i32) -> Result<i32, ArithmeticError> {
                a.checked_sub(b).ok_or(ArithmeticError::Overflow)
            }

            fn mul(a: i32, b: i32) -> Result<i32, ArithmeticError> {
                a.checked_mul(b).ok_or(ArithmeticError::Overflow)
            }

    .. compliant_example::
        :id: compl_ex_BgUHiRB4kc4b
        :status: draft

        Wrapping behavior must be explicitly requested. This compliant example uses wrapping functions.

        .. code-block:: rust

            fn add(a: i32, b: i32) -> i32 {
                a.wrapping_add(b)
            }

            fn sub(a: i32, b: i32) -> i32 {
                a.wrapping_sub(b)
            }

            fn mul(a: i32, b: i32) -> i32 {
                a.wrapping_mul(b)
            }

    .. compliant_example::
        :id: compl_ex_BhUHiRB4kc4b
        :status: draft

        Wrapping behavior call also be implemented using the ``Wrapping<T>`` type as in this compliant solution.

>         The ``Wrapping<T``> type is a ``struct`` found in the ``std::num`` module that explicitly enables two's complement wrapping arithmetic for the inner type ``T`` (which must be an integer or ``usize/isize``). 

Reviewer: ```suggestion
        The ``Wrapping<T>`` type is a ``struct`` found in the ``std::num`` module that explicitly enables two's complement wrapping arithmetic for the inner type ``T`` (which must be an integer or ``usize/isize``). 
```

Fixing a typo.

I'd also add links to the [docs](https://doc.rust-lang.org/stable/core/num/struct.Wrapping.html) because it would be very useful to have those at hand.

Author: I think you would either say:

``Wrapping<T>`` is a

or

The ``Wrapping<T>`` type is a

I think the second is a little better, and what we would use in the C Standard.

        The ``Wrapping<T>`` type provides a consistent way to force wrapping behavior in all build modes,
        which is useful in specific scenarios like implementing cryptography or hash functions where wrapping arithmetic is the intended behavior.

        .. code-block:: rust

            use std::num::Wrapping;

            fn add(si_a: Wrapping<i32>, si_b: Wrapping<i32>) -> Wrapping<i32> {
                si_a + si_b
            }

            fn sub(si_a: Wrapping<i32>, si_b: Wrapping<i32>) -> Wrapping<i32> {
                si_a - si_b
            }

            fn mul(si_a: Wrapping<i32>, si_b: Wrapping<i32>) -> Wrapping<i32> {
                si_a * si_b
            }

            fn main() {    
                let si_a = Wrapping(i32::MAX);
                let si_b = Wrapping(i32::MAX);

>                 println!("Adding {} by {} has a sum of {}", si_a, si_b, add(si_a, si_b))

Author: I half like the change.  I think "sums" is more specific than "results in" so I'll use that.

            }

    .. compliant_example::
        :id: compl_ex_BgUHiSB4kc4b
        :status: draft

        Saturation semantics means that any result that falls outside the valid range of the integer type is clamped to the maximum or minimum value instead wrapping around or resulting in an error.
        Saturation semantics always conform to this rule because they ensure that integer operations do not result in arithmetic overflow.

>         This compliant solution uses saturating functions to provide saturation semantics for some basic arithmetic operations.

Reviewer: ```suggestion
        Saturation semantics means that instead of wrapping around or resulting in an error,
        any result that falls outside the valid range of the integer type is clamped:

        - To the maximum value, if the result were to be greater than the maximum value, or
        - To the minimum value, if the result were to be smaller than the minimum,

        Saturation semantics always conform to this rule because they ensure that integer operations do not result in arithmetic overflow. 
        This compliant solution shows how to use saturating functions to provide saturation semantics for some basic arithmetic operations.
```

Added more clarity.

        .. code-block:: rust

            fn add(a: i32, b: i32) -> i32 {
                a.saturating_add(b)
            }

            fn sub(a: i32, b: i32) -> i32 {
                a.saturating_sub(b)
            }

            fn mul(a: i32, b: i32) -> i32 {
                a.saturating_mul(b)
            }

    .. compliant_example::
        :id: compl_ex_BgUHiSB4kd4b
        :status: draft

        ``Saturating<T>`` is a wrapper type in Rust’s standard library (``core::num::Saturating<T>``) that makes arithmetic operations on the wrapped value perform saturating arithmetic instead of wrapping, panicking, or overflowing.

>         The ``Saturating<T>`` is useful when you have a section of code or a data type where all arithmetic must be saturating.

Reviewer: ```suggestion
        ``Saturating<T>`` is useful when you have a section of code or a data type where all arithmetic must be saturating.
```

        This compliant solution uses the ``Saturating<T>`` type to define several functions that perform basic integer operations using saturation semantics.

        .. code-block:: rust

            use std::num::Saturating;

            fn add(si_a: Saturating<i32>, si_b: Saturating<i32>) -> Saturating<i32> {
                si_a + si_b
            }

            fn sub(si_a: Saturating<i32>, si_b: Saturating<i32>) -> Saturating<i32> {
                si_a - si_b
            }

            fn mul(si_a: Saturating<i32>, si_b: Saturating<i32>) -> Saturating<i32> {
                si_a * si_b
            }

            fn main() {    
                let si_a = Saturating(i32::MAX);
                let si_b = Saturating(i32::MAX);

>                 println!("Adding {} by {} has a sum of {}", si_a, si_b, add(si_a, si_b))

Reviewer: ```suggestion
                println!("Adding {} to {} results in {}", si_a, si_b, add(si_a, si_b))
```

Author: used math symbols

                if (si_b < 0 && si_a > INT_MAX + si_b)
                    || (si_b > 0 && si_a < INT_MIN + si_b)
                {
                   Err(ArithmeticError::Overflow)
                } else {
                   Ok(si_a - si_b)
                }
            }

	        fn mul(si_a: i32, si_b: i32) -> Result<i32, ArithmeticError> {
                if si_a == 0 || si_b == 0 {
                    return Ok(0);
                }

                // Detect overflow before performing multiplication
                if (si_a == -1 && si_b == INT_MIN) || (si_b == -1 && si_a == INT_MIN) {
                    Err(ArithmeticError::Overflow)
                } else if (si_a > 0 && (si_b > INT_MAX / si_a || si_b < INT_MIN / si_a))
                    || (si_a < 0 && (si_b > INT_MIN / si_a || si_b < INT_MAX / si_a))
                {
                    Err(ArithmeticError::Overflow)
                } else {
                    Ok(si_a * si_b)
                }
            }

    .. compliant_example::
        :id: compl_ex_BgUHiRB4kc4c
        :status: draft

        This compliant example uses safe checked addition instead of manual bounds checks.

>         Checked functions can reduce readibility when complex arithmetic expressions are needed.

Reviewer: ```suggestion
        Checked functions can reduce readability when complex arithmetic expressions are needed.
```

I'm suggesting this change real quick so that the build passes. Full review coming in a bit

    :scope: system
    :tags: security, performance, numerics

    Eliminate arithmetic overflow to avoid runtime panics and unexpected wraparound behavior.
    Use explicit wrapping or saturation semantics in cases where these behaviors are intentional.

>     Range checking can be accomplished using a variety of mechanisms, provided that the possibility of arithmetic overflow is elimianted.

Author: @felix91gr  I added the following two lines:

Arithmetic overflow will panic in debug mode, but wraparound in release mode.
Any wraparound behavior must be explicitly specified to ensure the same behavior in both debug and release modes.

>     Inline arithmetic expressions are often important for readability.

Reviewer: ```suggestion
```
This is correct, but it goes somewhere else. Perhaps in an Appendix to the Guidelines? Since it's not a guideline per se.

Anyway. The first bit of a Guideline is the Amplification section. This section is normative, and so every sentence in it is expected to be an actionable requirement of some sort.
